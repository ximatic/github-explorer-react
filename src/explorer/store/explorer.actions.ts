export enum ExplorerAction {
  Reset = 'explorer/reset',
  //ResetToken = "explorer/resetToken",
  VerifyToken = 'explorer/verifyToken',
  VerifyTokenSuccess = 'explorer/verifyTokenSuccess',
  VerifyTokenError = 'explorer/verifyTokenError',
  //Token = 'explorer/token',
  LoadRepositories = 'explorer/loadRepositories',
  LoadRepositoriesSuccess = 'explorer/loadRepositoriesSuccess',
  LoadRepositoriesError = 'explorer/loadRepositoriesError',
  LoadRepository = 'explorer/loadRepository',
  LoadRepositorySuccess = 'explorer/loadRepositorySuccess',
  LoadRepositoryError = 'explorer/loadRepositoryError',
  PageInfo = 'explorer/pageInfo',
}
