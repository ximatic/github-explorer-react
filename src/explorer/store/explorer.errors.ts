export enum ExplorerError {
  InvalidToken = 'EXPLORER_ERROR_INVALID_TOKEN',
}
