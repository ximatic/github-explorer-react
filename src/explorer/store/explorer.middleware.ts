import { Middleware } from '@reduxjs/toolkit';

import { ApolloQueryResult, gql } from '@apollo/client';

import { apolloClient } from '../../ApolloProvider';

import { GraphqlPageInfo, GraphqlRepositoriesData, GraphqlRepositoriesNode } from '../models/explorer-graphql.model';
import { ExplorerPageInfo, Repository } from '../models/explorer.model';

import { ExplorerAction } from './explorer.actions';
import { loadRepositoriesError, loadRepositoriesSuccess, pageInfo, verifyTokenError, verifyTokenSuccess } from './explorer.slice';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const explorerMiddleware: Middleware = (storeAPI: any) => (next: any) => (action: any) => {
  const result = next(action);

  const storageKey = 'gh-explorer-token';

  // token

  function storeToken(token: string): void {
    localStorage.setItem(storageKey, JSON.stringify(token));
  }

  function resetToken(): void {
    localStorage.removeItem(storageKey);
  }

  // page info

  function createPageInfo(itemsPerPage: number, pageInfo: GraphqlPageInfo): ExplorerPageInfo {
    return {
      itemsPerPage,
      cursorStart: pageInfo.startCursor,
      cursorEnd: pageInfo.endCursor,
      hasNextPage: pageInfo.hasNextPage,
      hasPreviousPage: pageInfo.hasPreviousPage,
    };
  }

  switch (action.type) {
    case ExplorerAction.Reset:
      resetToken();
      break;
    case ExplorerAction.VerifyToken:
      try {
        const token = action.payload.token;
        fetch('https://api.github.com/octocat', {
          headers: {
            Authorization: `Bearer ${token}`,
          },
        })
          .then(async (data: Response) => {
            if (!data.ok) {
              storeAPI.dispatch(verifyTokenError());
              return;
            }
            if (action.payload.storeToken) {
              storeToken(token);
            }
            storeAPI.dispatch(verifyTokenSuccess({ token }));
          })
          .catch(() => {
            storeAPI.dispatch(verifyTokenError());
          });
      } catch {
        storeAPI.dispatch(verifyTokenError());
      }
      break;
    case ExplorerAction.LoadRepositories:
      try {
        const pagination = action.payload.pagination;
        const token = storeAPI.getState().explorer.token;

        apolloClient
          .query({
            query: gql`
                query {
                  search(query: "is:public sort:stars-desc", type: REPOSITORY, ${pagination.itemsKey}: ${pagination.itemsValue}, ${pagination.cursorKey}: ${pagination.cursorValue}) {
                    nodes {
                      ...repositoryFragment
                    }
                    pageInfo {
                      endCursor
                      startCursor
                      hasNextPage
                      hasPreviousPage
                    }
                  }
                }
                fragment repositoryFragment on Repository {
                  name
                  createdAt
                  owner {
                    login
                  }
                  description
                  stargazerCount
                }
              `,
            context: {
              headers: {
                Authorization: `Bearer ${token}`,
              },
            },
          })
          .then((result: ApolloQueryResult<GraphqlRepositoriesData>) => {
            const repositories = result.data.search.nodes.map((node: GraphqlRepositoriesNode) => {
              return {
                name: node.name,
                owner: node.owner?.login || 'N/A',
                stars: node.stargazerCount,
                createdAt: node.createdAt,
                description: node.description,
              } as Repository;
            });

            storeAPI.dispatch(loadRepositoriesSuccess({ repositories }));
            storeAPI.dispatch(pageInfo({ pageInfo: createPageInfo(pagination.itemsValue, result.data.search.pageInfo) }));
          });
      } catch {
        storeAPI.dispatch(loadRepositoriesError());
      }
      break;
  }

  return result;
};
