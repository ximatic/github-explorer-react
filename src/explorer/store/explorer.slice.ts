import { createSlice, ReducerCreators, type PayloadAction } from '@reduxjs/toolkit';

import { ExplorerEvent, ExplorerEventName, ExplorerEventType, ExplorerState } from './explorer.state';
import { ExplorerPageInfo, ExplorerPagination, Repository } from '../models/explorer.model';

const initialState: ExplorerState = {
  token: null,
  repositories: null,
  repository: null,
  pageInfo: null,
  error: null,
  event: null,
};

export const explorerSlice = createSlice({
  name: 'explorer',
  initialState: initialState,
  reducers: (create: ReducerCreators<ExplorerState>) => ({
    // reset
    reset: create.reducer(() => initialState),
    // verify token
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    verifyToken: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string; storeToken: boolean }>) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Processing,
      };
    }),
    verifyTokenSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ token: string }>) => {
      state.token = action.payload.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Success,
      };
    }),
    verifyTokenError: create.reducer((state: ExplorerState) => {
      state.token = initialState.token;
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.VerifyToken,
        type: ExplorerEventType.Error,
      };
    }),
    // load repositories
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    loadRepositories: create.reducer((state: ExplorerState, action: PayloadAction<{ pagination: ExplorerPagination }>) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Processing,
      };
    }),
    loadRepositoriesSuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repositories: Repository[] }>) => {
      state.repositories = action.payload.repositories;
      state.repository = initialState.repository;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoriesError: create.reducer((state: ExplorerState) => {
      state.repositories = initialState.repositories;
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepositories,
        type: ExplorerEventType.Error,
      };
    }),
    // load repository
    loadRepository: create.reducer(
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      (state: ExplorerState, action: PayloadAction<{ owner: string; name: string; pagination: ExplorerPagination }>) => {
        state.pageInfo = initialState.pageInfo;
        state.event = {
          name: ExplorerEventName.LoadRepository,
          type: ExplorerEventType.Processing,
        };
      },
    ),
    loadRepositorySuccess: create.reducer((state: ExplorerState, action: PayloadAction<{ repository: Repository }>) => {
      state.repository = action.payload.repository;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Success,
      };
    }),
    loadRepositoryError: create.reducer((state: ExplorerState) => {
      state.repository = initialState.repository;
      state.pageInfo = initialState.pageInfo;
      state.event = {
        name: ExplorerEventName.LoadRepository,
        type: ExplorerEventType.Error,
      };
    }),
    // pagination
    pageInfo: create.reducer((state: ExplorerState, action: PayloadAction<{ pageInfo: ExplorerPageInfo }>) => {
      state.pageInfo = action.payload.pageInfo;
    }),
  }),
  selectors: {
    selectExplorerEvent: (state: ExplorerState): ExplorerEvent | null => state.event,
    selectToken: (state: ExplorerState): string | null => state.token,
    selectRepositories: (state: ExplorerState): Repository[] | null => state.repositories,
    selectRepository: (state: ExplorerState): Repository | null => state.repository,
    selectPageInfo: (state: ExplorerState): ExplorerPageInfo | null => state.pageInfo,
  },
});

export const {
  reset,
  verifyToken,
  verifyTokenSuccess,
  verifyTokenError,
  loadRepositories,
  loadRepositoriesSuccess,
  loadRepositoriesError,
  loadRepository,
  loadRepositorySuccess,
  loadRepositoryError,
  pageInfo,
} = explorerSlice.actions;

export const { selectExplorerEvent, selectToken, selectRepositories, selectRepository, selectPageInfo } = explorerSlice.selectors;
