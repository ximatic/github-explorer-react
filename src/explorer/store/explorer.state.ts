import { Repository, ExplorerPageInfo } from '../models/explorer.model';
import { ExplorerError } from './explorer.errors';

export enum ExplorerEventName {
  VerifyToken = 'verify-token',
  LoadRepositories = 'load-repositories',
  LoadRepository = 'load-repository',
}

export enum ExplorerEventType {
  Processing = 'processing',
  Success = 'success',
  Error = 'error',
}

export interface ExplorerEvent {
  name: ExplorerEventName;
  type: ExplorerEventType;
  message?: string;
}

export interface ExplorerState {
  token: string | null;
  repositories: Repository[] | null;
  repository: Repository | null;
  pageInfo: ExplorerPageInfo | null;
  error: ExplorerError | null;
  event: ExplorerEvent | null;
}
