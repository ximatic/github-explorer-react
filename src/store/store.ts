import { configureStore } from '@reduxjs/toolkit';

import { explorerSlice } from '../explorer/store/explorer.slice';
import { explorerMiddleware } from '../explorer/store/explorer.middleware';

// export const store = configureStore({
//   reducer: {
//     explorer: explorerSlice.reducer,
//   },
//   middleware: (getDefaultMiddleware) => {
//     return getDefaultMiddleware().concat(explorerMiddleware);
//   },
// });

// export type AppStore = typeof store;
// export type AppDispatch = typeof store.dispatch;
// export type RootState = ReturnType<typeof store.getState>;

export const makeStore = () => {
  return configureStore({
    reducer: {
      explorer: explorerSlice.reducer,
    },
    middleware: (getDefaultMiddleware) => {
      return getDefaultMiddleware().concat(explorerMiddleware);
    },
  });
};

// Infer the type of makeStore
export type AppStore = ReturnType<typeof makeStore>;
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<AppStore['getState']>;
export type AppDispatch = AppStore['dispatch'];
