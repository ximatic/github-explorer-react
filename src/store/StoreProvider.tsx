'use client';

import { useRef } from 'react';
import { Provider } from 'react-redux';

import { AppStore, makeStore } from './store';

interface StoreProviderProps {
  readonly children: React.ReactNode;
}

export default function StoreProvider({ children }: StoreProviderProps) {
  const storeRef = useRef<AppStore | null>(null);
  if (!storeRef.current) {
    // Create the store instance the first time this renders
    storeRef.current = makeStore();
  }

  return <Provider store={storeRef.current}>{children}</Provider>;
}
