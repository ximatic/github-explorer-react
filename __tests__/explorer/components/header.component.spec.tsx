import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { PrimeReactProvider } from 'primereact/api';

import StoreProvider from '../../../src/store/StoreProvider.tsx';
import * as reactRedux from '../../../src/store/hooks';

import { reset } from '../../../src/explorer/store/explorer.slice';

import { MOCK_TOKEN_1 } from '../../../__mocks__/explorer.mocks.ts';

import Header from '../../../src/explorer/components/header.component.tsx';

describe('Header', () => {
  const renderComponent = () =>
    render(
      <StoreProvider>
        <BrowserRouter>
          <PrimeReactProvider>
            <Header />
          </PrimeReactProvider>
        </BrowserRouter>
      </StoreProvider>,
    );

  afterEach(() => {
    jest.clearAllMocks();
  });

  // snapshot tests

  test('matches snapshot', () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  // unit tests

  test('renders with proper title', () => {
    renderComponent();

    expect(screen.getByText(/GitHub Explorer/i)).toBeInTheDocument();
  });

  test('calls reset and navigates to / when logout button is clicked', () => {
    const dispatchMock = jest.fn();
    const navigateMock = jest.fn();

    jest.spyOn(reactRedux, 'useAppSelector').mockReturnValue(MOCK_TOKEN_1);
    jest.spyOn(reactRedux, 'useAppDispatch').mockReturnValue(dispatchMock);
    jest.mock('react-router-dom', () => ({
      ...(jest.requireActual('react-router-dom') as any),
      useNavigate: () => navigateMock,
    }));

    renderComponent();

    const logoutBtn = screen.getByTestId('button-logout');
    fireEvent.click(logoutBtn);

    expect(dispatchMock).toHaveBeenCalledWith(reset());
    // TODO - issue with detecting navigation action - investigate in the future
    //expect(navigateMock).toHaveBeenCalledWith('/');
  });
});
