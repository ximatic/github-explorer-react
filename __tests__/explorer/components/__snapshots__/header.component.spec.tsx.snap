// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-toolbar p-component"
    data-pc-name="toolbar"
    data-pc-section="root"
    role="toolbar"
  >
    <div
      class="p-toolbar-group-start p-toolbar-group-left"
      data-pc-section="start"
    />
    <div
      class="p-toolbar-group-center"
      data-pc-section="center"
    >
      <div
        class="flex w-full"
      >
        <div
          class="grow text-center"
        >
          <span
            class="pi pi-github mr-2"
          />
          GitHub Explorer
        </div>
      </div>
    </div>
    <div
      class="p-toolbar-group-end p-toolbar-group-right"
      data-pc-section="end"
    />
  </div>
</DocumentFragment>
`;
